### PR TITLE
Fix disposing root X.509 certificate prematurely for OCSP stapling

### LINKED ISSUE #81964: Kestrel: Segfault when specifying default certificate on Ubuntu 20.04

### PR DESCRIPTION
**Pull request by vcsjones**
Fixes #81964

This fixes an issue where we were capturing the root certificate for OCSP stapling purposes, however would dispose of the certificate when we needed it. This resulted in a null `Handle`, which would be fed in to OpenSSL and cause a null deref in native code, crashing the .NET process. This scenario only occurred in two-element certificate chains. If there was an intermediate in the chain, the issue does not occur.

This PR addresses this in two ways. First, it stops disposing of the root certificate when it is needed. Some platforms and Linux situations don't need the root, and in that scenario, it gets disposed. The second is to actually check the `Handle` prior to handing it off to native code with an `Assert`. Additionally, in release configurations, this will simply exit the OCSP stapling and return a null response, and further prevent future staples for happening in that context.

Lastly, this adds tests for two-element chains. Prior to the fix, the tests would segfault as reported by the customer.

Contributes to #81964

### CODE REVIEW

**Comment by vcsjones** [comment]
@bartonjs this slightly deviates from our discussion earlier, and I'll attempt to explain why. First, with this approach we don't need to duplicate the certificate as we previously discussed. Instead, the platform decides if it cares to keep the root or not.

Second, we can't do `AddRootCertificate` earlier before the disposal because that's an instance member on `ctx` - and that needs the intermediates for the constructor.

File: src/libraries/Common/tests/System/Security/Cryptography/X509Certificates/CertificateAuthority.cs [hunk 1]
```diff
@@ -854,8 +854,8 @@ private enum CertStatus
                 rootAuthority = new CertificateAuthority(
                     rootCert,
                     rootDistributionViaHttp ? certUrl : null,
-                    issuerRevocationViaCrl ? cdpUrl : null,
-                    issuerRevocationViaOcsp ? ocspUrl : null);
+                    issuerRevocationViaCrl || (endEntityRevocationViaCrl && intermediateAuthorityCount == 0) ? cdpUrl : null,
```
Comment (vcsjones) [inline src/libraries/Common/tests/System/Security/Cryptography/X509Certificates/CertificateAuthority.cs:857]: In the case of no intermediates, the root and the issuer are the same CA, so we need to include OCSP info in the AIA.

File: src/libraries/System.Net.Security/tests/FunctionalTests/CertificateValidationRemoteServer.cs [hunk 2]
```diff
@@ -243,71 +248,80 @@ public Task ConnectWithRevocation_RemoteServer_StapledOcsp_FromWindows(X509Revoc
...
-            using (responder)
-            using (rootAuthority)
-            using (intermediateAuthority)
-            using (serverCert)
-            using (issuerCert)
-            using (rootCert)
-            await using (SslStream tlsClient = new SslStream(clientStream))
-            await using (SslStream tlsServer = new SslStream(serverStream))
+            try
             {
```
Comment (vcsjones) [inline src/libraries/System.Net.Security/tests/FunctionalTests/CertificateValidationRemoteServer.cs:252]: Mostly whitespace differences here. Easier to review ignoring whitespace.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.cs [hunk 3]
```diff
@@ -114,7 +119,7 @@ public static SslStreamCertificateContext Create(X509Certificate2 target, X509Ce
             return ctx;
         }
 
-        partial void AddRootCertificate(X509Certificate2? rootCertificate);
+        private partial void AddRootCertificate(X509Certificate2? rootCertificate);
```
Comment (bartonjs) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.cs:122]: We could get rid of the partials in the other implementations by changing this to have a ref bool and the call to be something like

```C#
bool transferred = false;
ctx.AddRootCertificate(root, ref transferred);

if (!transferred)
{
    root.Dispose();
}
```

And that returns the feature to a strict opt-in (and the if should const-fold away)

**Review by bartonjs — APPROVED** [review]
Left a suggestion to reduce the number of places we need to repeat the implementation of a partial method; otherwise LGTM.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.cs (hunk 3 above)
Comment (vcsjones) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.cs]: TIL `ref` works on partials. Makes sense.

**Review by bartonjs — APPROVED** [review]

**Comment by vcsjones** [comment]
I only ran the outerloop tests for linux since the test and changes only affect Linux. The failures in outerloop look irrelevant to me, as the two new test cases passed in all legs.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.cs [hunk 4]
```diff
@@ -109,12 +114,19 @@ public static SslStreamCertificateContext Create(X509Certificate2 target, X509Ce
...
             // unless this context was built "offline", or this came from the internal Create(X509Certificate2)
             ctx.SetNoOcspFetch(offline || noOcspFetch);
-            ctx.AddRootCertificate(root);
+
+            bool transferredOwnership = false;
+            ctx.AddRootCertificate(root, ref transferredOwnership);
+
+            if (!transferredOwnership)
+            {
+                root?.Dispose();
```
Comment (wfurt) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.cs:123]: should we dispose it in finalizer if we do not do it here?

Comment (vcsjones) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.cs:123]: I don't think so. The safe handle backing the certificate will have its own finalizer that'll clean up native resources.

**Review by bartonjs — APPROVED** [review]

**Pull request merged**